### PR TITLE
Save/Resume optimizer state, scheduler state, and epoch

### PR DESCRIPTION
**Pull request by mehdidc**
Save/Resume optimizer state, scheduler state, and epoch
Previously only the weights were saved, but for resuming we also need optimizer state, scheduler state, and epoch.

### CODE REVIEW

**Comment by rom1504** [comment]
Could you check whether that works with/without deepspeed ?

**Comment by mehdidc** [comment]
Sure, will do

**Comment by afiaka87** [comment]
@mehdidc 

I had success with running the code in amp O1 and zero stage 3 using this notebook:

https://colab.research.google.com/gist/afiaka87/bbac0038b213a82c067b4766cdf45e0d/ci_train_dalle_pytorch_wip.ipynb

@rom1504 and others are trying to automate this a bit; but stuff like this weeds out lots of bugs and gives more guarantees about PRs from new contributors etc.

If it's useful; feel free to change the branch/repo name to yours in there and see if the save/resume is working correctly.

**Comment by mehdidc** [comment]
Thanks a lot @afiaka87, this is perfect, will try it out. The notebook you made and CI in general will be super useful.

**Comment by afiaka87** [comment]
@mehdidc All running well? 

@lucidrains I think this may be ready for merge.

File: train_dalle.py [hunk 1]
```diff
@@ -179,7 +178,9 @@ def cp_path_to_dir(cp_path, tag):
         assert dalle_path.exists(), 'DALL-E model file does not exist'
     loaded_obj = torch.load(str(dalle_path), map_location='cpu')
 
-    dalle_params, vae_params, weights = loaded_obj['hparams'], loaded_obj['vae_params'], loaded_obj['weights']
+    dalle_params, vae_params, weights= loaded_obj['hparams'], loaded_obj['vae_params'], loaded_obj['weights']
```
Comment (afiaka87) [inline train_dalle.py:181]: Nitpicking - could we get a space before that `=` sign there?

File: train_dalle.py [hunk 2]
```diff
@@ -191,6 +192,7 @@ def cp_path_to_dir(cp_path, tag):
         **dalle_params
     )
     IMAGE_SIZE = vae.image_size
+    resume_epoch = loaded_obj.get('epoch', 0)
```
Comment (afiaka87) [inline train_dalle.py:195]: Oh wow this has bit me so many times on resume!

File: train_dalle.py [hunk 3]
```diff
@@ -295,6 +298,8 @@ def group_weight(model):
 # optimizer
 
 opt = Adam(get_trainable_params(dalle), lr=LEARNING_RATE)
+if RESUME and opt_state:
```
Comment (afiaka87) [inline train_dalle.py:301]: Till now I've had to use the DeepSpeed native optimizers to resume state - didn't know you could do this!

**Review by afiaka87 — APPROVED** [review]

**Comment by rom1504** [comment]
I think let's just confirm this runs ok if using deepspeed and if not using deepspeed
otherwise also lgtm

**Comment by afiaka87** [comment]
I'll spin it up in the notebook again.

File: train_dalle.py (hunk 1 above)
Comment (mehdidc) [reply train_dalle.py]: Thanks,  fixed

**Comment by afiaka87** [comment]
@mehdidc @lucidrains

Confirmed training works in deepspeed, confirmed resuming from deepspeed works, and resume from python with `dalle.pt`

Using https://colab.research.google.com/gist/afiaka87/bbac0038b213a82c067b4766cdf45e0d/ci_train_dalle_pytorch_wip.ipynb

**Comment by lucidrains** [comment]
thanks all!

**Comment by mehdidc** [comment]
Thanks @afiaka87 for testing it again!

**Pull request merged**